File: js/examples/runtime_control.js
const { Tracer } = require('../tracer.js');

async function doSomeWork(tracer, label) {
    tracer.span(`Step 1: ${label}`);
    await new Promise(resolve => setTimeout(resolve, 20));

    tracer.span(`Step 2: ${label}`);
    await new Promise(resolve => setTimeout(resolve, 30));

    tracer.span(`Step 3: ${label}`);
    await new Promise(resolve => setTimeout(resolve, 15));
}

async function main() {
    console.log('🔧 Runtime Control Examples');
    console.log('============================');

    // Example 1: Enable/Disable tracing
    console.log('\n1️⃣ Enable/Disable Tracing');
    console.log('--------------------------');

    const tracer1 = new Tracer('Runtime Control Demo');

    console.log('✅ Tracing enabled:');
    await doSomeWork(tracer1, 'Enabled');
    tracer1.end();

    console.log('\n❌ Tracing disabled:');
    tracer1.setEnabled(false);
    await doSomeWork(tracer1, 'Disabled'); // Won't collect data
    tracer1.end(); // Won't print anything

    // Example 2: Silent mode
    console.log('\n2️⃣ Silent Mode (Data Collection Only)');
    console.log('--------------------------------------');

    const tracer2 = new Tracer('Silent Mode Demo', {
        style: 'colorful'
    });

    tracer2.setSilent(true); // Collect data but don't print

    console.log('🔇 Silent mode - collecting data...');
    await doSomeWork(tracer2, 'Silent');
    tracer2.end(); // Won't print

    // Access collected data programmatically
    const measurements = tracer2.getMeasurements();
    const totalDuration = tracer2.getTotalDuration();

    console.log(`📊 Collected ${measurements.length} measurements`);
    console.log(`⏱️  Total duration: ${totalDuration}ms`);

    measurements.slice(0, -1).forEach((m, i) => { // Exclude "End"
        console.log(`   ${i + 1}. ${m.statement}: ${m.duration}ms`);
    });

    // Example 3: Dynamic style changes
    console.log('\n3️⃣ Dynamic Style Changes');
    console.log('-------------------------');

    let tracer3 = new Tracer('Style Change Demo');

    console.log('🎨 Starting with default style:');
    await doSomeWork(tracer3, 'Default Style');
    tracer3.end();

    // Change to colorful style
    tracer3 = new Tracer('Style Change Demo', {
        style: 'colorful'
    });

    console.log('\n🌈 Changed to colorful style:');
    await doSomeWork(tracer3, 'Colorful Style');
    tracer3.end();

    // Change to minimal style
    tracer3 = new Tracer('Style Change Demo');
    tracer3.setOutputStyle('minimal');

    console.log('\n📝 Changed to minimal style:');
    await doSomeWork(tracer3, 'Minimal Style');
    tracer3.end();

    // Example 4: Custom print conditions
    console.log('\n4️⃣ Custom Print Conditions');
    console.log('---------------------------');

    const tracer4 = new Tracer('Conditional Printing', {
        minTotalDuration: 50 // Only print if total > 50ms
    });

    console.log('⚡ Fast execution (won\'t print):');
    tracer4.span('Quick task');
    await new Promise(resolve => setTimeout(resolve, 10));
    tracer4.end(); // Won't print because < 50ms

    console.log('🐌 Slow execution (will print):');
    const tracer5 = new Tracer('Conditional Printing', {
        minTotalDuration: 50 // Only print if total > 50ms
    });

    await doSomeWork(tracer5, 'Slow enough'); // Will print because > 50ms
    tracer5.end();

    // Example 5: Data inspection without printing
    console.log('\n5️⃣ Data Inspection');
    console.log('-------------------');

    const tracer6 = new Tracer('Data Inspection', {
        silent: true // Silent to avoid printing
    });

    tracer6.span('Database connection');
    await new Promise(resolve => setTimeout(resolve, 25));

    tracer6.span('Query execution');
    await new Promise(resolve => setTimeout(resolve, 75));

    tracer6.span('Result processing');
    await new Promise(resolve => setTimeout(resolve, 40));

    tracer6.end();

    // Inspect collected data
    console.log(`📈 Total measurements: ${tracer6.getMeasurements().length}`);
    console.log(`⏱️  Total execution time: ${tracer6.getTotalDuration()}ms`);
    console.log(`🔧 Tracer enabled: ${tracer6.isEnabled()}`);
    console.log(`🔇 Tracer silent: ${tracer6.isSilent()}`);
    console.log(`🎨 Output style: ${tracer6.getOutputStyle()}`);
}

main().catch(console.error); 

File: go/examples/runtime_control.go
//go:build ignore
// +build ignore

package main

import (
	"fmt"
	"time"

	tracing "github.com/LeDuyViet/quicktrace/go"
)

func doSomeWork(tracer *tracing.Tracer, label string) {
	time.Sleep(20 * time.Millisecond)
	tracer.Span("Step 1: " + label)

	time.Sleep(30 * time.Millisecond)
	tracer.Span("Step 2: " + label)

	time.Sleep(15 * time.Millisecond)
	tracer.Span("Step 3: " + label)
}

func main() {
	fmt.Println("🔧 Runtime Control Examples")
	fmt.Println("============================")

	// Example 1: Enable/Disable tracing
	fmt.Println("\n1️⃣ Enable/Disable Tracing")
	fmt.Println("--------------------------")

	tracer1 := tracing.NewSimpleTracer("Runtime Control Demo")

	fmt.Println("✅ Tracing enabled:")
	doSomeWork(tracer1, "Enabled")
	tracer1.End()

	fmt.Println("\n❌ Tracing disabled:")
	tracer1.SetEnabled(false)
	doSomeWork(tracer1, "Disabled") // Won't collect data
	tracer1.End()                   // Won't print anything

	// Example 2: Silent mode
	fmt.Println("\n2️⃣ Silent Mode (Data Collection Only)")
	fmt.Println("--------------------------------------")

	tracer2 := tracing.NewSimpleTracer("Silent Mode Demo",
		tracing.WithOutputStyle(tracing.StyleColorful))

	tracer2.SetSilent(true) // Collect data but don't print

	fmt.Println("🔇 Silent mode - collecting data...")
	doSomeWork(tracer2, "Silent")
	tracer2.End() // Won't print

	// Access collected data programmatically
	measurements := tracer2.GetMeasurements()
	totalDuration := tracer2.GetTotalDuration()

	fmt.Printf("📊 Collected %d measurements\n", len(measurements))
	fmt.Printf("⏱️  Total duration: %v\n", totalDuration)

	for i, m := range measurements[:len(measurements)-1] { // Exclude "End"
		fmt.Printf("   %d. %s: %v\n", i+1, m.Statement, m.Duration)
	}

	// Example 3: Dynamic style changes
	fmt.Println("\n3️⃣ Dynamic Style Changes")
	fmt.Println("-------------------------")

	tracer3 := tracing.NewSimpleTracer("Style Change Demo")

	fmt.Println("🎨 Starting with default style:")
	doSomeWork(tracer3, "Default Style")
	tracer3.End()

	// Change to colorful style
	tracer3 = tracing.NewSimpleTracer("Style Change Demo",
		tracing.WithOutputStyle(tracing.StyleColorful))

	fmt.Println("\n🌈 Changed to colorful style:")
	doSomeWork(tracer3, "Colorful Style")
	tracer3.End()

	// Change to minimal style
	tracer3 = tracing.NewSimpleTracer("Style Change Demo")
	tracer3.SetOutputStyle(tracing.StyleMinimal)

	fmt.Println("\n📝 Changed to minimal style:")
	doSomeWork(tracer3, "Minimal Style")
	tracer3.End()

	// Example 4: Custom print conditions
	fmt.Println("\n4️⃣ Custom Print Conditions")
	fmt.Println("---------------------------")

	tracer4 := tracing.NewSimpleTracer("Conditional Printing")

	// Set custom condition: only print if total > 50ms
	tracer4.SetPrintCondition(func(t *tracing.Tracer) bool {
		return t.GetTotalDuration() > 50*time.Millisecond
	})

	fmt.Println("⚡ Fast execution (won't print):")
	time.Sleep(10 * time.Millisecond)
	tracer4.Span("Quick task")
	tracer4.End() // Won't print because < 50ms

	fmt.Println("🐌 Slow execution (will print):")
	tracer5 := tracing.NewSimpleTracer("Conditional Printing")
	tracer5.SetPrintCondition(func(t *tracing.Tracer) bool {
		return t.GetTotalDuration() > 50*time.Millisecond
	})

	doSomeWork(tracer5, "Slow enough") // Will print because > 50ms
	tracer5.End()

	// Example 5: Inspection without printing
	fmt.Println("\n5️⃣ Data Inspection")
	fmt.Println("-------------------")

	tracer6 := tracing.NewSimpleTracer("Data Inspection",
		tracing.WithSilent(true)) // Silent to avoid printing

	time.Sleep(25 * time.Millisecond)
	tracer6.Span("Database connection")

	time.Sleep(75 * time.Millisecond)
	tracer6.Span("Query execution")

	time.Sleep(40 * time.Millisecond)
	tracer6.Span("Result processing")

	tracer6.End()

	// Inspect collected data
	fmt.Printf("📈 Total measurements: %d\n", len(tracer6.GetMeasurements()))
	fmt.Printf("⏱️  Total execution time: %v\n", tracer6.GetTotalDuration())
	fmt.Printf("🔧 Tracer enabled: %v\n", tracer6.IsEnabled())
	fmt.Printf("🔇 Tracer silent: %v\n", tracer6.IsSilent())
	fmt.Printf("🎨 Output style: %d\n", tracer6.GetOutputStyle())
}
